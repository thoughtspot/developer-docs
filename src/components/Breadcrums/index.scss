@import '../../assets/styles/variables.scss';

.breadcrumsWrapper {
    padding: 20px 0 20px;
    
    .breadcrumb {
        margin: 0;
        list-style: none;
        list-style-type: none;
        padding-inline-start: 0px;
        padding-inline-end: 0px;
        li {
            display: inline;
            margin: 0;
            font-weight: 300;
            color: var(--breadcrums-font-color);
            font-size: $font-size-breadcrums;
        }

        li + li:before {
            padding: 0px 5px;
            content: '\003E';
        }

        li a {
            text-decoration: none;
            color: var(--breadcrums-font-color);
            font-size: $font-size-breadcrums;
            &:hover {
                color: var(--secondary-color);
            }
        }

    }

}

@media screen and (max-width: $mobile-resolution-max) {
       .breadcrumsWrapper{
           padding: 0 0 20px 0;
           font-size: $font-size-breadcrums-mobile;
           line-height: 14pt;
       }
}





