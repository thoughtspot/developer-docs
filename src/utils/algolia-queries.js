const { JSDOM } = require('jsdom');
const { htmlToText } = require('html-to-text');
const config = require('../configs/doc-configs');
const { getAlgoliaIndex } = require('../configs/algolia-search-config');

const getPathPrefix = () => {
    return 'docs';
};

const getPath = (path) =>
    getPathPrefix() ? `${path}/${getPathPrefix()}` : path;

const stripLinks = (text) => {
    if (text) {
        const re = /<a\s.*?href=[\"\'](.*?)[\"\']*?>(.*?)<\/a>/g;
        const str = text;
        const subst = '$2';
        const result = str.replace(re, subst);
        return result;
    }
    return '';
};

const getTextFromHtml = (html) => {
    const text = htmlToText(`${html}`, {
        selectors: [
            { selector: 'a', options: { ignoreHref: true } },
            { selector: 'h1', format: 'skip' },
            { selector: 'h2', format: 'skip' },
            { selector: 'h3', format: 'skip' },
        ],
    });
    return text;
};

const pageQuery = `
query {
  allAsciidoc(sort: { fields: [document___title], order: ASC }) {
      edges {
          node {
              id
              document {
                title
              }
              pageAttributes {
                  pageid
                  title
                  description
              }
              html       
          }
      }
  }
  allFile(filter: {sourceInstanceName: {eq: "htmlFiles"}}) {
    edges {
        node {
            id
            extension
            dir
            name
            relativePath
       
        }
    }
  }
}
`;

function splitStringIntoChunks(input, n) {
    const chunkSize = Math.ceil(input.length / n);
    const result = [];

    for (let i = 0; i < input.length; i += chunkSize) {
        const chunk = input.slice(i, i + chunkSize);
        result.push(chunk);
    }

    return result;
}

const pageToAlgoliaRecordForASCII = (ele, type, node) => {
    const pageid = node.pageAttributes.pageid;
    let sectionId;
    let sectionTitle;
    if (type === 'section') {
        sectionId = ele.querySelector('h2').id;
        sectionTitle = ele.querySelector('h2').innerHTML;
    } else {
        sectionId = type;
        sectionTitle = node.document.title;
    }

    const body = ele && getTextFromHtml(ele.innerHTML);
    const len = new TextEncoder().encode(body).length;

    const numberOfChunks = len / 8000 + 1;
    const chunks = splitStringIntoChunks(body, numberOfChunks);

    return chunks.map((chunk, i) => ({
        objectID: `${node.id + sectionId}_chunk_${i}`,
        sectionId,
        sectionTitle,
        body: chunk,
        pageid,
        type: 'ASCII',
        title: node.document.title,
        link: `/docs/${pageid}`,
    }));
};

const queries = [
    {
        query: pageQuery,
        transformer: ({ data }) => {
            const data1 = [
                ...data.allAsciidoc.edges
                    .filter(
                        (edge) =>
                            edge.node.pageAttributes.pageid &&
                            edge.node.pageAttributes.pageid !== 'nav',
                    )
                    .reduce((acc, edge) => {
                        const newDiv = new JSDOM(`<div>${edge.node.html}</div>`)
                            .window.document;
                        const preambleEle = newDiv.querySelector('#preamble');
                        const preamble = preambleEle
                            ? pageToAlgoliaRecordForASCII(
                                  preambleEle,
                                  'preamble',
                                  edge.node,
                              )
                            : null;
                        const sections = Array.prototype.map.call(
                            newDiv.querySelectorAll('.sect1'),
                            (sect) =>
                                pageToAlgoliaRecordForASCII(
                                    sect,
                                    'section',
                                    edge.node,
                                ),
                        );
                        if (preamble) {
                            acc = [...acc, ...preamble];
                        }
                        return [
                            ...acc,
                            ...sections.reduce(
                                (accR, e) => [...accR, ...e],
                                [],
                            ),
                        ];
                    }, []),
            ];

            return data1;
        },
        indexName: getAlgoliaIndex(),
        settings: {
            attributesToSnippet: ['body:10'],
        },
    },
];

module.exports = {
    queries,
    pageToAlgoliaRecordForASCII,
};
