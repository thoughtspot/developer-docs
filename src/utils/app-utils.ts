import 'url-search-params-polyfill';
import {
    NAV_PREFIX,
    PREVIEW_PREFIX,
    TS_HOST_PARAM,
    TS_ORIGIN_PARAM,
    TS_APP_ROOT_PARAM,
    TS_PAGE_ID_PARAM,
    DEFAULT_HOST,
    DEFAULT_PREVIEW_HOST,
    DEFAULT_APP_ROOT,
} from '../configs/doc-configs';

/**
 * Parse query string into json object. (Polyfill by 'url-search-params-polyfill' npm package)
 * @param {string} queryParamStr - query string from location.search
 * @returns {object} which contains query params 'key: value' pairs
 */
export const queryStringParser = (queryParamStr: string) => {
    const queryParamObj: { [key: string]: string } = {};

    const entries = new URLSearchParams(queryParamStr).entries();
    let navPrefix = '';
    let tsHostUrl = DEFAULT_PREVIEW_HOST;

    for (const [key, value] of entries) {
        queryParamObj[key] = value;
        if (key === TS_HOST_PARAM) {
            navPrefix += `${TS_HOST_PARAM}=${value}&`;
        }
        if (key === TS_ORIGIN_PARAM) {
            if (value) {
                navPrefix += `${TS_ORIGIN_PARAM}=${encodeURIComponent(value)}&`;
                tsHostUrl = removeTrailingSlash(value.split('#')[0]);
            }
        }
        if (key === TS_APP_ROOT_PARAM) {
            navPrefix += `${TS_APP_ROOT_PARAM}=${value}&`;
        }
    }

    // check required params and add default if value is not available
    queryParamObj[TS_HOST_PARAM] = removeTrailingSlash(
        queryParamObj[TS_HOST_PARAM] || DEFAULT_HOST,
    );
    queryParamObj[TS_APP_ROOT_PARAM] = removeTrailingSlash(
        queryParamObj[TS_APP_ROOT_PARAM] || DEFAULT_APP_ROOT,
    );

    // prepare and set 'Main Navigation' links URL prefix
    // navPrefix += TS_PAGE_ID_PARAM;
    // when using the / based system we dont need to prefix with pageid
    queryParamObj[NAV_PREFIX] = navPrefix;

    // prepare and set 'Preview in Playground' button URL prefix
    queryParamObj[
        PREVIEW_PREFIX
    ] = `${tsHostUrl}/#${queryParamObj[TS_APP_ROOT_PARAM]}`;

    return queryParamObj;
};

export const removeTrailingSlash = (url: string) => {
    return url.replace(/\/$/, '');
};

/**
 * Used to check if the docs portal open as public site or inside embed iframe
 * @param {string} queryParamStr - query string from location.search
 * @returns {boolean} true, if it is opened as public site otherwise false
 */
export const isPublicSite = (queryParamStr: string) => {
    const entries = new URLSearchParams(queryParamStr).entries();
    for (const [key, value] of entries) {
        if ([TS_HOST_PARAM, TS_ORIGIN_PARAM, TS_APP_ROOT_PARAM].includes(key)) {
            return false;
        }
    }
    return true;
};

export default queryStringParser;
