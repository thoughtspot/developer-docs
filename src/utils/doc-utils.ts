const buildJSON = (element) => {
    const parentElement = element?.children;
    const subObj: any = {
        name: parentElement[0]?.textContent?.trim(),
        href: parentElement[0]?.querySelector('a')?.getAttribute('href'),
    };

    if (parentElement?.length > 1 && parentElement[1]?.querySelector('ul')) {
        Array.from(parentElement[1].querySelector('ul').children).forEach(
            (eachElement) => {
                if (!subObj.children) {
                    subObj.children = [];
                }
                subObj.children.push(buildJSON(eachElement));
            },
        );
    }
    return subObj;
};

export const fetchChild = (html: string) => {
    if (typeof window === 'undefined') {
        return [];
    }
    const divElement = document.createElement('div');
    divElement.innerHTML = html;
    const data = Array.from(
        divElement.querySelectorAll('ul.navSection > li'),
    ).map((element) => buildJSON(element));

    return data;
};

// Not being used anywhere
const getParentHref = (current) => {
    if (current.href) {
        return current.href;
    }
    if (current?.children?.length > 1) {
        return current.children[0]?.href;
    }
    return null;
};

export const getBreadcrumsPath = (data: any, pageid?: string) => {
    if (!pageid) {
        return [];
    }

    return data.reduce((previous, current) => {
        if (
            current.href === `?pageid=${pageid}` ||
            current.href === `/${pageid}`
        ) {
            // To avoid having link for the same page we are setting href to null
            return [{ name: current.name, href: null }];
        }
        if (current.children) {
            const parentObj = [
                { name: current.name, href: getParentHref(current) },
            ];
            const childObj = getBreadcrumsPath(current.children, pageid);
            if (childObj.length) {
                return [...parentObj, ...childObj];
            }
        }
        if (previous.length) {
            return previous;
        }
        return [];
    }, []);
};

/**
 * Used to parse html by replace all passThroughKeys with corresponding replacement values.
 * @param {string} html - raw html content
 * @param {object} params - json object containing passThroughKey: replacementValue pairs
 * @returns {string} parsed html content
 */
export const passThroughHandler = (html: string, params: object) => {
    let parsedHtml = html;
    const paramKeys = Object.keys(params);
    if (!html && paramKeys.length === 0) return parsedHtml;

    const customPassThroughStart = '{{';
    const customPassThroughEnd = '}}';

    paramKeys.map((key) => {
        parsedHtml = parsedHtml.replace(
            new RegExp(
                `${customPassThroughStart}${key}${customPassThroughEnd}`,
                'g',
            ),
            params[key] || '',
        );
    });

    return parsedHtml;
};
