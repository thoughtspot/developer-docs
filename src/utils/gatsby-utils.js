const getTutorialLinkFromEdge = (edge) => {
    // Tutorials module pageids follow pattern {subdirectory}_{final_url_stub}
    // to give unique IDs in system but allow directory structure in URL
    const { pageid: pageId } = edge.node.pageAttributes;
    const { relativePath: relPath } = edge.node.parent;
    // One-level of subdirectory part of stub
    const relPathSplit = relPath.split('/');
    const pageIdSplit = pageId.split('__');
    let finalPageId = pageId;
    if (pageIdSplit.length > 1) {
        finalPageId = pageIdSplit[1];
    }

    let finalPath = `/tutorials/${finalPageId}`;
    if (relPathSplit.length > 1) {
        finalPath = `/tutorials/${relPathSplit[0]}/${finalPageId}`;
    }

    return finalPath;
};

const getDocLinkFromEdge = (edge) => {
    const { pageid: pageId } = edge.node.pageAttributes;
    const { sourceInstanceName: sourceName } = edge.node.parent;

    if (sourceName === 'tutorials') {
        return getTutorialLinkFromEdge(edge);
    }

    return `/${pageId}`;
};

module.exports = {
    getDocLinkFromEdge,
};
