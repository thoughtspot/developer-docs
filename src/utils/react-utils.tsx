import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { IconContext } from '@react-icons/all-files';

const getHTMLFromComponent = (icon: JSX.Element, iconClass?: string) => {
    return ReactDOMServer.renderToStaticMarkup(
        <IconContext.Provider value={{ className: `icon ${iconClass}` }}>
            {icon}
        </IconContext.Provider>,
    );
};

export { getHTMLFromComponent };
