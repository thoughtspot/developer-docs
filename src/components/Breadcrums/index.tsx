import React from 'react';
import { HOME_PAGE_ID } from '../../configs/doc-configs';
import { getBreadcrumsPath } from '../../utils/doc-utils';
import './index.scss';

type BreadcrumsProps = {
    pageid?: string;
    breadcrumsData: any;
};

const Breadcrums: React.FC<BreadcrumsProps> = (props: BreadcrumsProps) => {
    const breadcrums = getBreadcrumsPath(props.breadcrumsData, props.pageid);
    return (
        <>
            {breadcrums.length ? (
                <div
                    data-testid="breadcrumbWrapper"
                    className="breadcrumsWrapper"
                >
                    <ul className="breadcrumb">
                        <li>
                            <a href={`/docs/${HOME_PAGE_ID}`}>
                                Developer Guides
                            </a>
                        </li>
                        {breadcrums.map(({ name, href }) => (
                            <li key={`${name}-${href}`}>
                                {href ? <a href={href}>{name}</a> : name}
                            </li>
                        ))}
                    </ul>
                </div>
            ) : null}
        </>
    );
};

export default Breadcrums;
