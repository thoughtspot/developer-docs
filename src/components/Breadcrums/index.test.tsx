import React from 'react';
import { render } from "@testing-library/react";

import Breadcrums from "./index";
import { getBreadcrumsPath } from '../../utils/doc-utils';

jest.mock('../../utils/doc-utils', () => ({
    getBreadcrumsPath: jest.fn().mockReturnValue([
        {
            "name": "ThoughtSpot Embedded",
            "href": "?pageid=embed-analytics"
        },
        {
            "name": "ThoughtSpot Developer portal",
            "href": null
        }
    ]),
}))

describe('Breadcrums', () => {
    const pageId = 'spotdev-portal';
    const breadcrumsData = [
        {
            "name": "Home",
            "href": "?pageid=introduction"
        },
        {
            "name": "Whatâ€™s new",
            "href": "?pageid=whats-new"
        },
        {
            "name": "ThoughtSpot Embedded",
            "href": "?pageid=embed-analytics",
            "children": [
                {
                    "name": "Integration guidelines",
                    "href": "?pageid=integration-guidelines"
                },
                {
                    "name": "ThoughtSpot Developer portal",
                    "href": "?pageid=spotdev-portal"
                },
                {
                    "name": "Developer Playground",
                    "href": "?pageid=dev-playground"
                }
            ]
        },
        {
            "name": "Authentication and security",
            "href": "?pageid=auth-overview",
            "children": [
                {
                    "name": "Developer access",
                    "href": "?pageid=developer-access"
                },
                {
                    "name": "Security settings",
                    "href": "?pageid=security-settings"
                },
                {
                    "name": "SAML SSO authentication",
                    "href": "?pageid=saml-sso"
                },
                {
                    "name": "Trusted authentication",
                    "href": "?pageid=trusted-auth"
                },
                {
                    "name": "Custom domain configuration",
                    "href": "?pageid=custom-domain-config"
                },
                {
                    "name": "Browser settings for embedding",
                    "href": "?pageid=browser-settings"
                },
                {
                    "name": "Access control and data security",
                    "href": "?pageid=embed-object-access"
                }
            ]
        },
        {
            "name": "Visual Embed SDK",
            "href": "?pageid=visual-embed-sdk",
            "children": [
                {
                    "name": "Get started",
                    "href": "?pageid=getting-started"
                },
                {
                    "name": "Embed user authentication",
                    "href": "?pageid=embed-auth"
                },
                {
                    "name": "Embed search",
                    "href": "?pageid=search-embed"
                },
                {
                    "name": "Embed a pinboard",
                    "href": "?pageid=embed-pinboard"
                },
                {
                    "name": "Embed a visualization",
                    "href": "?pageid=embed-a-viz"
                },
                {
                    "name": "Embed full application",
                    "href": "?pageid=full-embed"
                },
                {
                    "name": "Disable or hide menu actions",
                    "href": "?pageid=action-config"
                },
                {
                    "name": "Visual Embed SDK Reference",
                    "href": "?pageid=js-reference"
                }
            ]
        },
        {
            "name": "Custom actions",
            "href": "?pageid=custom-action-intro",
            "children": [
                {
                    "name": "Custom actions page",
                    "href": "?pageid=customize-actions"
                },
                {
                    "name": "Configure a URL action",
                    "href": "?pageid=custom-action-url"
                },
                {
                    "name": "Configure a callback action",
                    "href": "?pageid=custom-action-callback"
                },
                {
                    "name": "Add a custom action to a visualization",
                    "href": "?pageid=add-action-viz"
                },
                {
                    "name": "Add custom actions to a worksheet",
                    "href": "?pageid=add-action-worksheet"
                },
                {
                    "name": "Callback custom action example",
                    "href": "?pageid=push-data"
                },
                {
                    "name": "Custom action response payload",
                    "href": "?pageid=custom-action-payload"
                }
            ]
        },
        {
            "name": "Customization and rebranding",
            "href": "?pageid=customization-intro",
            "children": [
                {
                    "name": "Customize layout and styles",
                    "href": "?pageid=customize-style"
                },
                {
                    "name": "Customize links",
                    "href": "?pageid=customize-links"
                },
                {
                    "name": "Customize onboarding settings",
                    "href": "?pageid=customize-emails"
                }
            ]
        },
        {
            "name": "Runtime filters",
            "href": "?pageid=runtime-filters"
        },
        {
            "name": "REST API fundamentals",
            "href": "?pageid=rest-apis",
            "children": [
                {
                    "name": "Get started",
                    "href": "?pageid=rest-api-getstarted"
                },
                {
                    "name": "Authentication",
                    "href": "?pageid=api-auth-session"
                },
                {
                    "name": "Manage users and user groups",
                    "href": "?pageid=api-user-management"
                },
                {
                    "name": "Embed data using REST APIs",
                    "href": "?pageid=embed-data-restapi"
                },
                {
                    "name": "Paginate API response",
                    "href": "?pageid=rest-api-pagination"
                }
            ]
        },
        {
            "name": "REST API Reference",
            "href": "?pageid=rest-api-reference",
            "children": [
                {
                    "name": "User APIs",
                    "href": "?pageid=user-api"
                },
                {
                    "name": "Group APIs",
                    "href": "?pageid=group-api"
                },
                {
                    "name": "Session APIs",
                    "href": "?pageid=session-api"
                },
                {
                    "name": "Data connection APIs",
                    "href": "?pageid=connection-api"
                },
                {
                    "name": "Metadata APIs",
                    "href": "?pageid=metadata-api"
                },
                {
                    "name": "Admin APIs",
                    "href": "?pageid=admin-api"
                },
                {
                    "name": "TML APIs",
                    "href": "?pageid=tml-api"
                },
                {
                    "name": "Dependent objects APIs",
                    "href": "?pageid=dependent-objects-api"
                },
                {
                    "name": "Search data API",
                    "href": "?pageid=search-data-api"
                },
                {
                    "name": "Pinboard data API",
                    "href": "?pageid=pinboard-api"
                },
                {
                    "name": "Pinboard Export API",
                    "href": "?pageid=pinboard-export-api"
                },
                {
                    "name": "Security APIs",
                    "href": "?pageid=security-api"
                },
                {
                    "name": "Log streaming service API",
                    "href": "?pageid=logs-api"
                }
            ]
        },
        {
            "name": "Code samples",
            "href": "?pageid=code-samples"
        },
        {
            "name": "Resources",
            "children": [
                {
                    "name": "Playground",
                    "href": "https://try-everywhere.thoughtspot.cloud/v2/#/everywhere/playground/search"
                },
                {
                    "name": "REST API Explorer (Requires login)",
                    "href": "https://try-everywhere.thoughtspot.cloud/external/swagger"
                },
                {
                    "name": "ThoughtSpot Developers",
                    "href": "https://developers.thoughtspot.com"
                },
                {
                    "name": "Community",
                    "href": "https://community.thoughtspot.com/customers/s/"
                },
                {
                    "name": "Product Documentation",
                    "href": "https://cloud-docs.thoughtspot.com"
                }
            ]
        }
    ]
    it('should render correctly and show breadcrumbs', async () => {
        const { container, queryByTestId } = await render(<Breadcrums pageid={pageId} breadcrumsData={breadcrumsData} />);
        expect(container).toMatchSnapshot();
        expect(queryByTestId('breadcrumbWrapper')).toBeInTheDocument();
        expect(getBreadcrumsPath).toHaveBeenCalledTimes(1);
    });

})