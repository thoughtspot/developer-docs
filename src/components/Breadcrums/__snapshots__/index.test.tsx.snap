// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrums should render correctly and show breadcrumbs 1`] = `
<div>
  <div
    class="breadcrumsWrapper"
    data-testid="breadcrumbWrapper"
  >
    <ul
      class="breadcrumb"
    >
      <li>
        <a
          href="/docs/introduction"
        >
          Developer Guides
        </a>
      </li>
      <li>
        <a
          href="?pageid=embed-analytics"
        >
          ThoughtSpot Embedded
        </a>
      </li>
      <li>
        ThoughtSpot Developer portal
      </li>
    </ul>
  </div>
</div>
`;
